cosmwasm/contracts/bindings_factory/src/contract.rs

```diff
@@ -129,31 +129,60 @@ mod execute {
 
         }
 
-        // Convert the bech32 string back to 'Addr' type before passing to the canine_bindings helper API
+        // Convert the bech32 string back to 'Addr' type before passing to the canine_bindings helper API.
         let error_msg: String = String::from("Bindings contract address is not a valid bech32 address. Conversion back to addr failed");
         let bindings_contract = BindingsContract::new(deps.api.addr_validate(&bindings_address).expect(&error_msg));
         
-        // Execute the bindings contract with given msg
-        let cosmos_msg = bindings_contract.execute(msg, info.funds)?;
+        // Clone msg first for later use. 
+        let msg_clone = msg.clone();
 
-        // We only add the factory_cosmos_msg if it's non empty--i.e., we actually need it for creating a bindings contract 
+        // Prepare the bindings msg for broadcast below. 
+        let cosmos_msg = bindings_contract.execute(msg, info.funds)?;
 
         let mut messages: Vec<CosmosMsg> = Vec::new();
 
+        // Iff the factory_cosmos_msg--an instantiate variant--has a non zero code id, it will be added
+        // to the above 'messages' for broadcast.
+        // To figure this out:
+        // If 'factory_cosmos_msg' contains a WasmMsg::Instantiate2 variant,
+        // we extract the code_id from it and assign it to the above 'id'. 
         let mut id: u64 = 0;
 
         if let CosmosMsg::Wasm(wasm_msg) = factory_cosmos_msg.clone() {
            if let WasmMsg::Instantiate2 { admin: _, code_id, label: _, msg: _, funds: _, salt: _ } = wasm_msg {
                 id = code_id;
            }
         }
-
         if id != 0 {
-            messages.push(factory_cosmos_msg);
+            messages.push(factory_cosmos_msg); // This only happens if it's the user's first time using the evm outpost 
+        }
+
+        // If there is a collision, DO NOT broadcast.
+        let binary_msg: Binary = to_json_binary(&msg_clone).expect("Failed to convert msg to Binary");
+        let hashed_msg: [u8; 32] = hash_msg(&binary_msg);
+        let hashed_msg_hex = hash_to_hex(hashed_msg);
+
+        let mut collision: bool = false;
+
+        if let Some(value) = BROADCASTED_MSGS.may_load(deps.storage, (&evm_address, hashed_msg_hex.clone()))? {
```

Comment:
what is the time complexity of this? are we okay to keep saving hashes?

Reply:
O(1) - should be good